### PR TITLE
[BEAM-22] Improve ParDoEvaluator Factoring

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
This moves shared code into a common location.

Clone DoFn instances before constructing the DoFnRunner to
 avoid races.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @dhalperi

File: sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/ParDoSingleEvaluatorFactory.java [hunk 1]
```diff
@@ -43,7 +39,7 @@
       CommittedBundle<?> inputBundle,
       InProcessEvaluationContext evaluationContext) {
     @SuppressWarnings({"cast", "unchecked", "rawtypes"})
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/ParDoSingleEvaluatorFactory.java]: You dropped the cast.. are these all still needed?

Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/ParDoSingleEvaluatorFactory.java]: The cast is not, the remaining two still do suppress warnings. Done.

**Comment by dhalperi** [comment]
LGTM

**Comment by dhalperi** [comment]
@tgroh : please backport

**Pull request closed without merging**